Mortar deuce

lindsayad opened the pull request:
This is a new implementation of mortar, replacing the old implementation. Some results and comparison with the old implementation can be found at #13080.

Defining characteristics of the new `MortarConstraint`:
- Works in parallel; with `RelationshipManger` system we are able to get the proper ghosting across the interface
- Automatically generates a mortar segment mesh using the `AutomaticMeshGeneration` object whose segments will consist of contributions from at most one slave element and one master element, e.g. the mortar segment element will be contained within a slave element and the mortar segment element nodes are guaranteed to project onto the same master element
- Integral approximation by quadrature is done carefully for projection of master contributions onto the mortar segment to avoid conservation loss; convergence properties are consistent with theory across many discretizations: P2P1, P2P0, P1P1, P1P0. Again, see #13080 

There is still more work to be done for mortar (list over at #13080), but this PR is already quite large as it is and represents a good check-point in the road.

All the concept work is the product of @jwpeterson and is imported from his libMesh code

lindsayad committed:
Initial mortar work. Things compile

lindsayad committed:
Do some things to RealMortarConstraint

lindsayad committed:
Start working on RealMortarConstraint residual. Compiles

lindsayad committed:
Everything is into computeResidual. Need to split out into computeJacobian as well

lindsayad committed:
Split out into computeResidual and computeJacobian

lindsayad committed:
Add subdomain_id key which also adds the slave and master subdomains to AMG. (These should both be created using LowerDBlocksFromSideset)

lindsayad committed:
Mortar problem running but giving incorrect results

lindsayad committed:
Fix node projection methods to be correctly acting on subdomain pairs

lindsayad committed:
Debugging helper methods

lindsayad committed:
Add real mortar constraint test

lindsayad committed:
Clean up RealMortarConstraint. Remove all libMesh::out

lindsayad committed:
Ensure that the mortar mesh is generated before we determine the elements we need to ghost

lindsayad committed:
Make our mortar example work in parallel

lindsayad committed:
Generalize RealMortarConstraint

lindsayad committed:
Create GapHeatConductanceTest object

lindsayad committed:
Add fine gap conductance test

lindsayad committed:
Reorganize and add mortar-soln-continuity test. Refs #617

lindsayad commented:
@tophmatthews @bwspenc

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/304080/) on 2901c2f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13056/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13056/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 89905e95ae8845deb3a8efffdb4f8385aab90fb4`

lindsayad committed:
clang-format

lindsayad committed:
Delete unused

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/304150/) on c802642 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13056/site)

This comment will be updated on new commits.

lindsayad committed:
Generalize computation of nodal normals

bwspenc commented on framework/include/constraints/RealMortarConstraint.h:
```diff
@@ -0,0 +1,185 @@
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef REALMORTARCONSTRAINT_H
```
What's 'Real' about this?  Is it Real as in really being a mortar constraint, or is this referring to Real numbers?

lindsayad replied:
It's really a mortar constraint. This name will change; it will likely just replace the existing `MortarConstraint` class

lindsayad committed:
Infrastructure to add Edge3s

lindsayad committed:
Lay down interior nodes for EDGE3 elements

lindsayad committed:
Exclude interior nodes from eligible nearest neighbors in Nanoflann

lindsayad committed:
soln-continuity test sucess

lindsayad committed:
Don't try to do any ghosting activity with the AMG object until we can retrieve it

During initial mesh setup for DistributedMesh there is a call through to
ghosting functors. Our ghosting requires the existence of an AutomaticMortarGeneration
object, which is held by the FEProblemBase, which hasn't been created by the time of
this initial call. So if we don't have an executioner yet (which is actually setup
after the FEProleBase), then we just return and don't do anything

lindsayad committed:
Delete what I believe is unneccessary ghosting

lindsayad committed:
Add doc for FunctionGradientNeumannBC

moosebuild commented:
Job [Clean Machine](https://civet.inl.gov/job/305292/) on 892eb4b : invalidated by @permcody

lindsayad committed:
Restrict mortar to replicated mesh for now

lindsayad committed:
Migrate RealMortarConstraint to have same API as MortarConstraint

lindsayad committed:
Rip out the original MortarConstraint

This commit compiles but the old tests that use MortarConstraint derived
objects fail because the input parameters are different. I was thinking
about going and re-doing these old meshes to create the subdomain and
boundary ID objects that the new constraint needs, but this caused me to
realize that the old mortar system works in 3D. I feel that we can't rip
out an object that supports higher-D mortar than the new implementation, so
I'm abandoning this tear-out for now.

lindsayad committed:
SQA requirements

lindsayad committed:
Don't create too mortar ghosting functors for two constraints on the same mortar interface

lindsayad committed:
Adapt old mortar equalgradient test to new mortar
Adapt old mortar equalgradient test to new mortar

This will require passing the -partial option to the Exodiff binary
because we now have a different number of element blocks, elements,
and nodes. However the topography of the higher dimensional meshes
is the same so the exodiff tester works on the primal variable

lindsayad committed:
Add -partial option to Exodiff python tester

lindsayad committed:
Convert over other conforming first order tests

lindsayad committed:
Remove non-conforming 2d mortar tests

It's simply too much of a headache to try and convert over these old
tests. The Exodiff complains about non-matching nodes

lindsayad committed:
Convert 2nd order conforming 2d test

lindsayad committed:
Remove 3d mortar tests

lindsayad committed:
Clean up new tests

- Move new tests from constraints/mortar* into mortar/
- Remove SolutionContinuityTest because EqualValueConstraint does the same thing

lindsayad committed:
Adapt GapHeatConductance to new mortar constraint

lindsayad commented:
@YaqiWang Do you guys do 3D mortar in rattlesnake?

lindsayad committed:
More framework scrubbing of old mortar

lindsayad committed:
Allow new mortar system to do periodic conditions/constraints

lindsayad committed:
User can now supply subdomain and boundary names to MortarConstraint instead of ids

lindsayad committed:
Adapt combined mortar example to new mortar

lindsayad committed:
Enable running mortar on a displaced mesh

This commit was done to enable running the combined gap_heat_transfer_mortar_displaced
test with the new mortar implementation. As was true for the old mortar implementation
this only works for auxiliary displacement variables. E.g. we are not going to update
the mortar segment meshes during residual and jacobian evaluations yet, so we cannot
use non-linear displacement variables. Additionally the AugmentSparsityOnInterface
ghosting functor assumes that the element ids are equivalent between reference and
displaced meshes. This will certainly be violated at times when performing AMR, e.g
this probably won't work with AMR.

lindsayad committed:
Finish documentation

lindsayad commented:
@idaholab/moose-team, @bwspenc I believe this is ready for review

lindsayad commented:
Don't be scared off by the line count, at least 2,000 lines are from a single `.msh` file... 😄

YaqiWang commented:
Sorry, I did not notice the tagging because of the heavy traffic in this PR. Rattlesnake is using mortar only for constraints with Lagrange multipliers. It does not have tests with mis-aligned meshes. It does not have 3D tests either. We switched to penalty method quickly after realizing the drawbacks of the old mortar method long ago. If this PR is in place, it will be very nice to finish the mortar application in Rattlesnake for radiation transport.

lindsayad committed:
Replace libmesh with moose method naming style

lindsayad committed:
Enable dynamic regeneration of the mortar mesh

We now execute mortar mesh generation on residual evaluations when
the MortarData object has displaced AutomaticMortarGeneration objects.
We also added a test using auxiliary displacement variables that are functions
of time

lindsayad committed:
Ghost the whole mortar interface when mesh is dynamic

lindsayad committed:
Change key for master_node_and_elem_to_xi1_slave_elem

Previously the key was a tuple of <const Node *, const Elem*>. This can
lead to different ordering of the map contents across different processes
which leads to a different sequence of element generation for the mortar
segment mesh because we iterate over this map when projecting master nodes
onto the slave face and splitting the mortar segement elements. The new
key is a tuple of <dof_id_type, const Node *, const Elem *> so that there
will be consistent ordering across processes.

lindsayad committed:
Overhaul mortar

- Move fe and variable reinit into a separate functor
- Make lagrange multiplier variable optional
- Mechanical contact test

lindsayad committed:
We need to be able to reinit precisely how we need to for mortar

lindsayad committed:
Update to new relationship manager system

permcody commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -0,0 +1,252 @@
+
+  // Map from nodes to connected lower-dimensional elements on the slave/master subdomains.
+  std::map<dof_id_type, std::vector<const Elem *>> nodes_to_slave_elem_map;
+  std::map<dof_id_type, std::vector<const Elem *>> nodes_to_master_elem_map;
```
These should probably be `unordered_map`s

permcody commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -0,0 +1,252 @@
+  // (Master Node, Master Elem) -> (xi^(1), Slave Elem)
+  // where they are inverse-projected along the nodal normal
+  // direction.
+  std::map<std::pair<const Node *, const Elem *>, std::pair<Real, const Elem *>>
```
`unordered_map` here too (and above)

permcody commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -0,0 +1,252 @@
+  std::unordered_multimap<dof_id_type, dof_id_type> mortar_interface_coupling;
+
+  // Container for storing the nodal normal vector associated with each slave node.
+  std::unordered_map<const Node *, Point> slave_node_to_nodal_normal;
```
OK - so you used `unordered_map` here. I don't immediately see why we can't in the other places I've indicated, am I missing something?

lindsayad replied:
K, switched all maps that could be switched to `unordered`

permcody commented on framework/include/constraints/NanoflannMeshAdaptor.h:
```diff
@@ -0,0 +1,209 @@
+    // Loop over the elements of the Mesh, for those in the requested
+    // subdomain, add its node ids to the _legal_point_indices set.
+    for (MeshBase::const_element_iterator el = _mesh.active_elements_begin(),
+                                          end_el = _mesh.active_elements_end();
```
You should probably use the range_based version here.

permcody commented on framework/include/constraints/NanoflannMeshAdaptor.h:
```diff
@@ -0,0 +1,209 @@
+      const Elem * elem = *el;
+      if (elem->subdomain_id() == _sid)
+      {
+        for (unsigned int n = 0; n < elem->n_vertices(); ++n)
```
MooseIndex

permcody commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -0,0 +1,1187 @@
+        "Must specify slave and master boundary ids before building node-to-elem maps.");
+
+  // Construct nodes_to_slave_elem_map
+  for (MeshBase::const_element_iterator el = mesh.elements_begin(), end_el = mesh.elements_end();
```
I don't think we have a range based version, but you can use the `as_range` helper to make one here.

permcody commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -0,0 +1,1187 @@
+    if (!this->slave_boundary_subdomain_ids.count(slave_elem->subdomain_id()))
+      continue;
+
+    for (unsigned int n = 0; n < slave_elem->n_vertices(); ++n)
```
MooseIndex

permcody commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -0,0 +1,1187 @@
+
+    for (unsigned int n = 0; n < slave_elem->n_vertices(); ++n)
+    {
+      std::vector<const Elem *> & vec = nodes_to_slave_elem_map[slave_elem->node_id(n)];
```
RHS map[] access? You should probably assert unless getting a new entry is OK and desired here.

lindsayad replied:
Yes, it's desired

permcody commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -0,0 +1,1187 @@
+
+    // Make sure we found one.
+    if (current_mortar_segment == libmesh_nullptr)
+      libmesh_error_msg("Unable to find appropriate mortar segment during linear search!");
```
`nullptr` and `mooseError`

permcody commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -0,0 +1,1187 @@
+
+    // Get the set of master_node neighbors.
+    const std::vector<const Elem *> & master_node_neighbors =
+        this->nodes_to_master_elem_map.at(master_node->id());
```
`.at()` is nasty since it throws something that we usually don't catch and also doesn't always play well with MPI_Abort. You are better with a `find` and `mooseAssert`.

permcody commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -0,0 +1,1187 @@
+
+    // Sanity check
+    if (master_node_neighbors.size() == 0 || master_node_neighbors.size() > 2)
+      libmesh_error_msg("We must have either 1 or 2 master side nodal neighbors, but we had "
```
mooseError()

permcody commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -0,0 +1,1187 @@
+    // swap them if that's not the case.
+    const Elem * left_master_elem = master_node_neighbors[0];
+    const Elem * right_master_elem =
+        (master_node_neighbors.size() == 2) ? master_node_neighbors[1] : libmesh_nullptr;
```
`nullptr`

permcody commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -0,0 +1,1187 @@
+
+    // Storage for z-component of cross products for determining
+    // orientation.
+    std::vector<Real> slave_node_cps(2), master_node_cps(2);
```
Possibly use `std::array`

permcody commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -0,0 +1,1187 @@
+  FEType nnx_fe_type(FIRST, LAGRANGE);
+
+  // Build FE objects from those types.
+  UniquePtr<FEBase> nnx_fe_face(FEBase::build(dim, nnx_fe_type));
```
std::unique_ptr()

permcody commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -0,0 +1,1187 @@
+      Point nodal_normal = this->slave_node_to_nodal_normal.at(slave_node);
+
+      // Data structure for performing Nanoflann searches.
+      Real query_pt[3] = {(*slave_node)(0), (*slave_node)(1), (*slave_node)(2)};
```
std::array

permcody commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -0,0 +1,1187 @@
+
+      if (!projection_succeeded)
+      {
+        libMesh::out << "Failed to find master Elem into which slave node "
```
mooseWarning()? or mooseInfo()? Not sure if this comes out all the time or not.

permcody commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -0,0 +1,1187 @@
+  kd_tree.buildIndex();
+
+  for (MeshBase::const_element_iterator el = mesh.active_elements_begin(),
+                                        end_el = mesh.active_elements_end();
```
range-based for

permcody commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -0,0 +1,1187 @@
+
+      if (!projection_succeeded)
+      {
+        libMesh::out << "Failed to find point from which master node "
```
Another libMesh::out

permcody commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -0,0 +1,1187 @@
+{
+  // Must call compute_nodal_normals() first!
+  if (slave_node_to_nodal_normal.empty())
+    libmesh_error_msg("No entries found in the slave node -> nodal normal map.");
```
mooseError

permcody commented on framework/src/constraints/MortarConstraint.C:
```diff
@@ -15,243 +15,442 @@
+    params.addParam<BoundaryID>("master_boundary_id", "The id of the master boundary sideset.");
+    params.addParam<BoundaryID>("slave_boundary_id", "The id of the slave boundary sideset.");
+    params.addParam<SubdomainID>("master_subdomain_id", "The id of the master subdomain.");
+    params.addParam<SubdomainID>("slave_subdomain_id", "The id of the slave subdomain.");
```
In general, we always read in "Name"s instead of "ID"s. The reason is that "Name"s handle ID cases seamlessly.

permcody commented on framework/src/constraints/MortarConstraint.C:
```diff
@@ -15,243 +15,442 @@
+    _master_subdomain_id(
+        isParamValid("master_subdomain_id")
+            ? getParam<SubdomainID>("master_subdomain_id")
+            : _mesh.getSubdomainID(getParam<SubdomainName>("master_subdomain_name"))),
```
Oh yeah... Definitely don't do this. Name's will handle either names or IDs

permcody commented on framework/src/constraints/MortarConstraint.C:
```diff
@@ -15,243 +15,442 @@
+  for (MeshBase::const_element_iterator
+           el = _amg.mortar_segment_mesh.active_local_elements_begin(),
+           end_el = _amg.mortar_segment_mesh.active_local_elements_end();
+       el != end_el;
```
range-based for

permcody commented on framework/src/constraints/MortarConstraint.C:
```diff
@@ -15,243 +15,442 @@
+    _master_primal_offset = _slave_primal_offset + _dof_indices_slave_interior_primal.size();
+
+    // These only get initialized if there is a master Elem associated to this segment.
+    const Elem * master_ip = libmesh_nullptr;
```
`nullptr`

permcody commented on framework/src/constraints/MortarData.C:
```diff
@@ -0,0 +1,75 @@
+
+    ////////////////////////////////////////////////////////////////////////////////
+    // Build the mortar segment mesh on the slave boundary.
+    ////////////////////////////////////////////////////////////////////////////////
```
Looks like John style comments :)

permcody commented on framework/src/constraints/MortarSegmentInfo.C:
```diff
@@ -0,0 +1,128 @@
+MortarSegmentInfo::print() const
+{
+  libMesh::out << "xi^(1)_a=" << xi1_a << ", xi^(1)_b=" << xi1_b << std::endl;
+  libMesh::out << "xi^(2)_a=" << xi2_a << ", xi^(2)_b=" << xi2_b << std::endl;
```
_console

permcody reviewed (dismissed):
I hate to hold this up, but there are still a lot of libMesh-ish constructs in some of the classes we inherited from John. If you want to address that in a follow-up PR, I'm ok with that. It might be nice to clean all of those things up first. AutomaticMortarGeneration is particularly full of non-MOOSE-like code.

lindsayad committed:
Address review

lindsayad committed:
Create MooseHashing header

- Move the hash_combine function from TheWarehouse into MooseHashing.h
  and put it in the Moose namespace.
- Specialize std::hash for std::pair

lindsayad committed:
Finish addressing review

lindsayad commented:
Alright the do not merge label is for real this time. I don't think this can go in as is because it's too disruptive to Rattlesnake. I've been working towards adding a third MOOSE reinit--elem, neighbor, and now ***lower_d***--that should enable updating Rattlesnake more easily. I was planning for that to be in a follow-on PR because this one is already huge, but what can you do when you're responsible for making sure you never break anything downstream.

YaqiWang commented:
@lindsayad Maybe I can help if you have more info? Is it caused by constraints, splitting mesh or something else?

lindsayad committed:
Changes for Rattlesnake

lindsayad commented:
It's in `reinit` and `reinitSide` for `PNSNConstraint`. The problem is that in this PR there is no good way to reinit on the lower dimensional element through the `Assembly` object; the lagrange multiplier, primal slave, and primal master variables are reinited manually in the `MortarConstraint` class directly through libMesh. This is fine as long as those are the only variables you use. But you are coupling in other variables in that class, so the `MortarConstraint` class in this PR is simply not flexible enough. The nice thing about the code I've written as a follow-on is that you will not need to override any methods like `reinit` and `reinitSide`. You will be able to grab coupled variables using traditional means like `coupledValue`.

YaqiWang commented:
I see. This sounds something we need eventually. It is hard to make the implementation flexible to meet all needs. I am glad Rattlesnake provides an useful requirement ;-)

lindsayad committed:
Compile checkpoint

lindsayad committed:
Add SubProblem::reinitLowerDElemRef phase 1

Adds:
- Assembly::reinitLowerDElemRef and needed callees
- SystemBase::prepareLowerD and needed callees
- Assembly::prepareLowerD and needed callees

lindsayad committed:
SubProblem::reinitLowerDElemRef phase 2

lindsayad committed:
Overhaul MooseVariableFE to remove code duplication

This also introduces a lot more const'ness in MOOSE. Making things
semantically const also requires introduction of some `mutable` keywords

lindsayad committed:
Get simple diffusion to run

- Make _ad*grad_phi_data mutable
- Make dofIndices in MooseVariableBase virtual and override in MooseVariableFE
- Remove non-const getter in MooseVariableBase for dof indices as it breaks
  encapsulation
- Update places that used the non-const getter to use the const getter (make
  use of universal references that ensure maintaining any cv qualifications
  fomr the method return type)
- Compile with gcc
    - Implement undefined refernces (MooseVariableData::duDotDu,
      getVector, getMatrix)
    - Remove leaks from fe objects and qrules in Assembly
    - in ComputeMortarFunctor store constraints as const MortarConstraintBase *
    - Always put default at end of switch case

lindsayad committed:
All moose tests running successfully other than mortar

- Override numberOfDofs for MooseVariableData
- Fix neighbor qrule
- Calculate correct ad_offset in MooseVariableData::computeAD based on _element_type
- Protect against null element when doing MooseVariableData::prepareAux

lindsayad committed:
Finish the redesign of the mortar redesign

lindsayad committed:
Add simple test for analytic verification

lindsayad committed:
Adapt modules to new const overhaul

Also:
- Fix ghosting for mortar on displaced meshes (loop over all active elements)
- Remove ridiculous HeatConductionBC for the gap_heat_transfer_mortar_displaced test

lindsayad committed:
Fixes for unity build

lindsayad committed:
Update to use pragma once

lindsayad committed:
Shuffle some using declarations

lindsayad committed:
Loosen up the MortarConstraint interface

lindsayad commented:
@YaqiWang How am I supposed to debug your insane actions when I can't even use `DumpObjectsProblem`? Result of attempt:
```
A SN scheme requires Problem/type to be SNFEProblem
```
I need to be able to disable individual kernels and constraints to test some of the Jacobians, because right now the Jacobian is terribly wrong and I can't get a couple of input files to converge. From first inspection the lagrange multiplier variable Jacobians look correct, but I can't go any deeper without some fine-grained control. I don't really feel like going through `MultischemeAction` commenting out individual pieces of code.

lindsayad committed:
Create a non-AD mortar constraint

This is necessary because we cannot convert over to dynamic derivative storage
because at the present time, it is too slow. Because of that, Rattlesnake
cannot use any AD objects. Hence we need a non-AD version of the constraint.

permcody commented on framework/src/relationshipmanagers/AugmentSparsityOnInterface.C:
```diff
@@ -0,0 +1,152 @@
+    _has_attached_amg = true;
+  }
+
+  const CouplingMatrix * const null_mat = libmesh_nullptr;
```
libmesh_nullptr -> nullptr

permcody commented on framework/src/relationshipmanagers/AugmentSparsityOnInterface.C:
```diff
@@ -0,0 +1,152 @@
+          coupled_elements.insert(std::make_pair(elem, null_mat));
+        auto ip = elem->interior_parent();
+        if (ip->processor_id() != p)
+          coupled_elements.insert(std::make_pair(ip, null_mat));
```
maybe emplace()?

permcody commented on framework/src/relationshipmanagers/AugmentSparsityOnInterface.C:
```diff
@@ -0,0 +1,152 @@
+        // list of Elems coupled to us.
+        const Elem * cross_interface_neighbor_ip = cross_interface_neighbor->interior_parent();
+        if (cross_interface_neighbor_ip && cross_interface_neighbor_ip->processor_id() != p)
+          coupled_elements.insert(std::make_pair(cross_interface_neighbor_ip, null_mat));
```
emplace()

permcody reviewed (approved):
This is an awesome piece of work. I don't see any reason to hold it up any longer. I have a couple of nitpicks but I don't see any design-level issues with this PR.
